### DISCUSSION TITLE
Specify OpenAI Execution settings per Invoke

### DISCUSSION THREAD

**Discussion by LefanTan** [microsoft/semantic-kernel#5797] [Q&A] [answered]
Hi all! 

Currently, we have an internal tool that needs a way to allow users to specify their preferred GPT model  (as well as other configs such as Temperature). 

It seems like in SK, the Models are pre-defined in `        builder.AddOpenAIChatCompletion(_appSettings.App.SemanticKernel.TextModel, chatCompletionClient);
` as well as `config.json`. I'm currently using `kernel.InvokeAsync` but may be using a planner soon too.  How can I approach this?

**Comment by dmytrostruk** [comment, marked as answer]
Hi @LefanTan , it's possible to set OpenAI execution settings per `Invoke` by passing settings model within `KernelArguments`. Here is the syntax:
https://github.com/microsoft/semantic-kernel/blob/9b8a218f5b6df9dcc72d69d989aff9b904cdecf0/dotnet/samples/KernelSyntaxExamples/Example58_ConfigureExecutionSettings.cs#L47-L54

Please let me know if that resolves your scenario. Thank you!

**Reply by LefanTan** [reply]
This is perfect, thank you!

**Reply by LefanTan** [reply]
@dmytrostruk Upon further inspection, it seems like the ModelId defined during `builder.AddOpenAIChatCompletion(modelId)`
will override whatever specified in `KernelArguments`

Example:
![image](https://github.com/microsoft/semantic-kernel/assets/12474777/5ba05e6b-737e-481d-824a-f9a7c47d3682)

**Reply by dmytrostruk** [reply]
@LefanTan That's because you need to register multiple instances of OpenAI connector per model, like in this example:

```csharp
// Build a kernel with multiple chat completion services
var builder = Kernel.CreateBuilder()
    .AddOpenAIChatCompletion(
        modelId: "gpt-4-0125-preview",
        apiKey: TestConfiguration.OpenAI.ApiKey)
    .AddOpenAIChatCompletion(
        modelId: "gpt-3.5-turbo-0125",
        apiKey: TestConfiguration.OpenAI.ApiKey);

Kernel kernel = builder.Build();

var prompt = "Hello AI, what can you do for me?";

var result = await kernel.InvokePromptAsync(prompt, new(new OpenAIPromptExecutionSettings { ModelId = "gpt-3.5-turbo-0125" }));
```

In this case, it will use the model specified in `OpenAIPromptExecutionSettings.ModelId`. 

It is also possible to inject your own `IAIServiceSelector` implementation, which will select AI service based on your conditions. Here is an example with custom AI service selector:
https://github.com/microsoft/semantic-kernel/blob/57b55af5786ed44630896b5ad4916636c3deba8e/dotnet/samples/KernelSyntaxExamples/Example62_CustomAIServiceSelector.cs
